src/domain/usecases/data-entry/amc/utils/calculationConsumptionProductLevelData.ts

```diff
@@ -48,13 +48,16 @@ export function calculateConsumptionProductLevelData(
         period,
     });
 
-    const dddCombinations: DDDCombinationsData[] = atcClassification.find(({ name }) => name === DDD_COMBINATIONS_NAME)
-        ?.data as DDDCombinationsData[];
-    const dddData: DDDData[] = atcClassification.find(({ name }) => name === DDD_NAME)?.data as DDDData[];
-    const dddAlterations: DDDAlterationsData[] = atcClassification.find(({ name }) => name === DDD_ALTERATIONS_NAME)
-        ?.data as DDDAlterationsData[];
-    const conversionFactorData: ConversionFactorData[] = atcClassification.find(({ name }) => name === CONVERSION_NAME)
-        ?.data as ConversionFactorData[];
+    const dddCombinations = atcClassification.find(({ name }) => name === DDD_COMBINATIONS_NAME)?.data as
+        | DDDCombinationsData[]
```

Comment:
@anagperal  What is the advantage of defining 
export type GlassATCVersion = Array<
    GlassATCVersionData<
        DDDCombinationsData | ConversionFactorData | DDDData | ATCData | DDDAlterationsData | ATCAlterationsData
    >
>; 
instead of its actual structure which is 
type GlassATCVersion = {
    atc: ATCData[];
    ddd_combinations: DDDCombinationsData[];
    conversion: ConversionFactorData[];
    ddd_alterations: DDDAlterationsData[];
    atc_alterations: ATCAlterationsData[];
}[];
The current type declaration does not auto suggest object properties when coding, requires extra code to parse the object and necessitates forcing types with "as" keyword which leads to issue such as this one. 
What is the benefit of this way of type declaration ?

Reply:
@9sneha-n I've used this way of type but we can use also these options:

```typescript
// Option 1
export type GlassATCVersionData = {
    name: "atc" | "ddd_combinations" | "ddd" | "conversion" | "ddd_alterations" | "atc_alterations";
    data: (DDDCombinationsData | ConversionFactorData | DDDData | ATCData | DDDAlterationsData | ATCAlterationsData)[];
};
export type GlassATCVersion = Array<GlassATCVersionData>;


// Option 2
export type ATC = {
    name: "atc";
    data: ATCData[];
};
export type DDDComb = {
    name: "ddd_combinations";
    data: DDDCombinationsData[];
};
export type DDD = {
    name: "ddd";
    data: DDDData[];
};
export type Conver = {
    name: "conversion";
    data: ConversionFactorData[];
};
export type DDDAlt = {
    name: "ddd_alterations";
    data: DDDAlterationsData[];
};
export type ATCAlt = {
    name: "atc_alterations";
    data: ATCAlterationsData[];
};
export type GlassATCVersionData = ATC | DDDComb | DDD | Conver | DDDAlt | ATCAlt;
export type GlassATCVersion = Array<ATC | DDDComb | DDD | Conver | DDDAlt | ATCAlt>;
```
But in everything you have to say what type of data you are using to see the auto suggestions.

Also, the type you propose would not be exactly the structure of ATC version, but maybe it would be another option. We could make a mapper in GlassATCDefaultRepository to return it like this instead of an array of {name: "name", data: []} and use this type. What do you think?

Reply:
Hi @anagperal , this was very interesting and i researched a little bit. I think our answer is typescript "Narrowing - Using type predicates" . Here is a link to the article i read https://www.typescriptlang.org/docs/handbook/2/narrowing.html#using-type-predicates

I did a quick test with the Option2 you provided, and it works. But i think this same concept should work with your original code as well (creating a is<Type> function in the find)

![image](https://github.com/EyeSeeTea/glass-dev/assets/83749675/b8d5c681-a660-4596-b998-7b1d996b0d8a)

Do you think this would work ?

Reply:
Also @anagperal  a very cheeky question, BUT, since we are storing the atc data in our datastore, do you think we could change the format? (make the type as the key)

Reply:
Hi @9sneha-n ! Thanks for all this info, I didn't know about the "Narrowing - Using type predicates". From my point of view, I agree it's better change the types in GLASS code for having the structure of ATC looks like and object with the keys. For the Datastore structure I looked at the one for the glass modules, I don't know whether we have any advantage in having one structure or the other in the Datastore. I think that we have two options:
1. Talk with @deeonwuli and change https://github.com/EyeSeeTea/d2-reports/pull/132, to do a mapper and upload this object structure to Datastore
2. We keep that in Datastore and we do mapper in GLASS code
And in both we have to change the GLASS code (no problem, would be quick and easy)
In my opinion, any of them are valid, the important thing is to be coordinated in the change.